File: server/py_server/grps_framework/model_infer/torch_inferer.py
# Copyright 2022 netease. All rights reserved.
# Author zhaochaochao@corp.netease.com
# Date   2022/9/13
# Brief  Torch deep learning model inferer. Including model load and model infer.
import os
import re

import torch

from grps_framework.apis.grps_pb2 import GrpsMessage
from grps_framework.converter.torch_tensor_converter import TorchTensorConverter
from grps_framework.logger.logger import logger
from grps_framework.model_infer.inferer import ModelInferer


class TorchModelInferer(ModelInferer):
    """Torch model(torch script model format) inferer."""

    def __init__(self):
        super().__init__()
        self._model = None
        self._inp_device = None
        self._converter = TorchTensorConverter()  # used when no converter mode

    def init(self, path, device=None, args=None):
        """
        Initiate model inferer class with model path and device.

        Args:
            path: Model path, it can be a file path or a directory path.
            device: Device to run model.
            args: More args.

        Raises:
            Exception: If init failed, raise exception. Will be caught by server and show error message to user when
            start service.
        """
        if torch.__version__ < '1.8.0':
            raise RuntimeError(
                'PyTorch version should be greater than or equal 1.8.0. Current version is {}.'.format(
                    torch.__version__))
        super().init(path, device, args)

        if not device:
            raise ValueError('Device not specified.')
        elif device == 'cpu':
            self._device = 'cpu'
            self._inp_device = 'cpu'
        elif device in ['cuda', 'gpu']:
            # == cuda:0
            self._device = 'cuda:0'
            self._inp_device = 'cuda:0'
        elif re.match('^cuda:\\d+$', device):
            self._inp_device = self._device
        elif re.match('^gpu:\\d+$', device):
            self._device = 'cuda:' + self._device[4:]
            self._inp_device = self._device
        elif device[:8] == 'original':
            self._device = 'original'
            self._inp_device = device[9:]
            if self._inp_device not in ['cpu', 'cuda', 'gpu'] and \
                    not re.match('^cuda:\\d+$', self._inp_device) and \
                    not re.match('^gpu:\\d+$', self._inp_device):
                raise ValueError('Invalid inp_device: {}, should be cpu, cuda, gpu, cuda:[num] or gpu:[num].'
                                 .format(self._inp_device))
            if re.match('^gpu:\\d+$', self._inp_device):
                self._inp_device = 'cuda:' + self._inp_device[4:]
        else:
            raise ValueError(
                'Invalid device: {}, should be cpu, cuda, gpu, cuda:[num], gpu:[num] or original.'.format(self._device))

        if args is not None:
            self._customized_op_paths = args.get('customized_op_paths')
            if self._customized_op_paths is not None:
                if type(self._customized_op_paths) is not list:
                    raise ValueError(
                        'Invalid customized_op_paths: {}, should be list.'.format(self._customized_op_paths))
                for path in self._customized_op_paths:
                    if not os.path.exists(path):
                        raise ValueError('Customized op path not exists: {}.'.format(path))
                    torch.ops.load_library(path)
                    logger.info('Load customized op library: {}.'.format(path))

    def load(self):
        """
        Load model from model path.

        Returns:
            True if load model successfully, otherwise False.

        Raises:
            Exception: If load failed, raise exception and exception will be caught by server and show error message
            to user when start service.
        """
        logger.info('Load model from {} to device {}.'.format(self._path, self._device))
        self._model = torch.jit.load(self._path)
        if self._device != 'original':
            self._model.to(self._device)
        if not self._model:
            return False
        self._model.eval()
        logger.info('Load model successfully.')
        return True

    def infer(self, tensors, context):
        """
        The inference function is used to make a prediction call on the given input request.

        Args:
            context: grps context
            tensors: Model infer input, which is output of converter preprocess function. When in `no converter mode`,
            will skip converter preprocess and directly use GrpsMessage as input.

        Returns:
            Model infer output, which will be input of converter postprocess function. When in `no converter mode`, it
            will skip converter postprocess and should directly use GrpsMessage as output.

        Raises:
            Exception: If infer failed, raise exception and exception will be caught by server and return error
            message to client.
        """
        with torch.no_grad():
            if str(type(tensors)) == str(GrpsMessage):
                # no converter mode
                tensors = self._converter.preprocess(tensors, context)
                tensors = self.infer(tensors, context)
                return self._converter.postprocess(tensors, context)

            # place tensors to device
            if isinstance(tensors, torch.Tensor):
                tensors = tensors.to(self._inp_device)
                return self._model(tensors)
            elif isinstance(tensors, list) or isinstance(tensors, tuple):
                if isinstance(tensors, tuple):
                    tensors = list(tensors)
                for idx, tensor in enumerate(tensors):
                    if isinstance(tensor, torch.Tensor):
                        tensors[idx] = tensor.to(self._inp_device)
                return self._model(*tensors)
            elif isinstance(tensors, dict):
                for key, value in tensors.items():
                    if isinstance(value, torch.Tensor):
                        tensors[key] = value.to(self._inp_device)
                return self._model(**tensors)
            else:
                return self._model(tensors)

    def batch_infer(self, inp, contexts: list):
        """
        Batch infer.

        Args:
            inp: Model infer input, which is output of converter batch_preprocess function. When in `no converter mode`,
            will skip converter batch_preprocess and directly use GrpsMessage list as input.
            contexts: Grps context list.

        Returns:
            Model infer output, which will be input of converter batch_postprocess function. When in `no converter mode`,
            it will skip converter batch_postprocess and should directly use GrpsMessage list as output.

        Raises:
            Exception: If batch infer failed, can raise exception and exception will be caught by server and return
            error message to client.
        """
        if type(inp) is list and str(type(inp[0])) == str(GrpsMessage):
            # No converter mode.
            tensors = self._converter.batch_preprocess(inp, contexts)
            tensors = self.batch_infer(tensors, contexts)
            return self._converter.batch_postprocess(tensors, contexts)
        else:
            return self.infer(inp, contexts[0])  # contexts are useless.

File: server/cpp_server/src/model_infer/torch_inferer.cc
/*
 * Copyright 2022 netease. All rights reserved.
 * Author zhaochaochao@corp.netease.com
 * Date   2023/02/08
 * Brief  torch model inferer implementation.
 */

#include "torch_inferer.h"

#include <ATen/core/Tensor.h>
#include <dlfcn.h>
#include <torch/csrc/api/include/torch/version.h>

#include <regex>

#include "logger/logger.h"

namespace netease::grps {
void TorchModelInferer::Init(const std::string& path, const std::string& device, const YAML::Node& args) {
  ModelInferer::Init(path, device, args);
  if (device == "cpu") {
    torch_dev_ = torch::kCPU;
    inp_dev_ = torch::kCPU;
  } else if (device == "cuda" || device == "gpu") {
    // == cuda:0
    device_ = "cuda:0";
    torch_dev_ = torch::Device("cuda:0");
    inp_dev_ = torch::Device("cuda:0");
  } else if (std::regex_match(device, std::regex("^cuda:\\d+$"))) { // device: cuda:0, cuda:1, ...
    torch_dev_ = torch::Device(device);
    inp_dev_ = torch::Device(device);
  } else if (std::regex_match(device, std::regex("^gpu:\\d+$"))) {
    device_ = "cuda:" + device.substr(4);
    torch_dev_ = torch::Device("cuda:" + device.substr(4));
    inp_dev_ = torch::Device("cuda:" + device.substr(4));
  } else if (device.substr(0, 8) == "original") { // original device specified when exported model.
    device_ = "original";
    std::string inp_device = device.substr(9);
    if (inp_device != "cpu" && inp_device != "cuda" && inp_device != "gpu" &&
        !std::regex_match(inp_device, std::regex("^cuda:\\d+$")) &&
        !std::regex_match(inp_device, std::regex("^gpu:\\d+$"))) {
      LOG4(ERROR, "Invalid inp_device: " << inp_device << ", should be cpu, cuda, gpu, cuda:[num] or gpu:[num].");
      throw InfererException("Invalid inp_device: " + inp_device +
                             ", should be cpu, cuda, gpu, cuda:[num] or gpu:[num]");
    }
    if (std::regex_match(inp_device, std::regex("^gpu:\\d+$"))) {
      inp_device = "cuda:" + inp_device.substr(4);
    }
    inp_dev_ = torch::Device(inp_device);
  } else {
    LOG4(ERROR, "Init torch script model inferer failed, device: "
                  << device << " is not supported, should be cpu, cuda, gpu, cuda:[num], gpu:[num] or original.");
    throw InfererException("Device is not supported, device: " + device);
  }

  if (args && !args.IsNull() && args.IsMap()) {
    // Load customized op lib if exists.
    auto customized_op_paths = args["customized_op_paths"];
    if (customized_op_paths && !customized_op_paths.IsNull() && customized_op_paths.IsSequence()) {
      for (const auto& op_path : customized_op_paths) {
        if (op_path && !op_path.IsNull() && op_path.IsScalar()) {
          auto path_str = op_path.as<std::string>();
          if (!path_str.empty()) {
            auto* handler = dlopen(path_str.c_str(), RTLD_LAZY);
            if (handler == nullptr) {
              std::string err_str = "Load customized op lib failed, path: ";
              err_str.append(path_str).append(", error: ").append(dlerror());
              LOG4(ERROR, err_str);
              throw InfererException(err_str);
            }
            LOG4(INFO, "Load customized op lib success, path: " << path_str);
          }
        }
      }
    }
  }
}

void TorchModelInferer::Load() {
  try {
    if (device_ == "original") {
      module_ = torch::jit::load(path_);
    } else {
      module_ = torch::jit::load(path_, torch_dev_);
    }
    module_.eval();

    auto inputs = module_.get_method("forward").function().getSchema().arguments();
    for (int i = 0; i < inputs.size(); i++) {
      if (i == 0) {
        continue; // Skip self.
      }
      inp_tensor_names_.emplace_back(inputs[i].name());
    }
    if (inp_tensor_names_.empty()) {
      LOG4(ERROR, "Load torch script model inferer failed, no input tensor.");
      throw InfererException("No input tensor.");
    }

    auto outputs = module_.get_method("forward").function().getSchema().returns();
    for (const auto& output : outputs) {
      out_tensor_names_.emplace_back(output.name());
    }
    if (out_tensor_names_.empty()) {
      LOG4(ERROR, "Load torch script model inferer failed, no output tensor.");
      throw InfererException("No output tensor.");
    }
  } catch (const c10::Error& e) {
    LOG4(ERROR, "Load torch script model inferer failed, error: " << e.what());
    throw InfererException(e.what());
  }
  LOG4(INFO, "Load torch script model inferer success, model inferer path: "
               << path_ << ", device: " << device_ << ", " << "input tensor count: " << inp_tensor_names_.size()
               << ", input tensor names: " << inp_tensor_names_ << ", output tensor count: " << out_tensor_names_.size()
               << ", output tensor names: " << out_tensor_names_);
}

void TorchModelInferer::Infer(const std::vector<std::pair<std::string, TensorWrapper>>& inputs,
                              std::vector<std::pair<std::string, TensorWrapper>>& outputs,
                              GrpsContext& context) {
#if TORCH_VERSION_MAJOR > 1 || (TORCH_VERSION_MAJOR == 1 && TORCH_VERSION_MINOR >= 9)
  c10::InferenceMode inference_mode_guard(infer_mode_);
#endif
  torch::NoGradGuard no_grad_guard;

  // Prepare input tensor.
  if (inputs.empty()) {
    throw InfererException("No input tensor.");
  }
  std::vector<torch::jit::IValue> input_tensor;
  for (int i = 0; i < inp_tensor_names_.size(); ++i) {
    if (i >= inputs.size()) {
      throw InfererException("Input tensor is not enough, required: " + std::to_string(inp_tensor_names_.size()) +
                             ", actual: " + std::to_string(inputs.size()));
    }
    if (inputs[i].first.empty()) { // Use default name and sequence.
      if (inputs[i].second.torch_tensor == nullptr) {
        throw InfererException("Input tensor is null, index: " + std::to_string(i));
      }
      input_tensor.emplace_back(inputs[i].second.torch_tensor->to(inp_dev_));
    } else { // Find by name.
      auto iter = std::find_if(inputs.begin(), inputs.end(),
                               [name = inp_tensor_names_[i]](const auto& item) { return item.first == name; });
      if (iter == inputs.end()) {
        throw InfererException("Input tensor is not found, name: " + inp_tensor_names_[i]);
      }
      if (iter->second.torch_tensor == nullptr) {
        throw InfererException("Input tensor is null, name: " + inp_tensor_names_[i]);
      }
      input_tensor.emplace_back(iter->second.torch_tensor->to(inp_dev_));
    }
  }
  try {
    outputs.clear();
    auto output = module_(input_tensor);
    if (output.isTensor()) {
      outputs.emplace_back(out_tensor_names_[0], std::move(output.toTensor()));
    } else if (output.isTuple()) {
      auto tuple = output.toTuple();
      for (size_t i = 0; i < tuple->elements().size(); ++i) {
        const std::string& name = (i < out_tensor_names_.size() && !out_tensor_names_[i].empty())
                                    ? out_tensor_names_[i]
                                    : "output_" + std::to_string(i);
        outputs.emplace_back(name, std::move(const_cast<c10::IValue&>(tuple->elements()[i]).toTensor()));
      }
    } else if (output.isTensorList()) {
      auto tensor_list = output.toTensorList();
      for (size_t i = 0; i < tensor_list.size(); ++i) {
        const std::string& name = (i < out_tensor_names_.size() && !out_tensor_names_[i].empty())
                                    ? out_tensor_names_[i]
                                    : "output_" + std::to_string(i);
        outputs.emplace_back(name, std::move(tensor_list.get(i)));
      }
    } else if (output.isGenericDict()) {
      auto dict = output.toGenericDict();
      for (const auto& item : dict) {
        outputs.emplace_back(item.key().toString()->string(), std::move(item.value().toTensor()));
      }
    } else {
      LOG4(ERROR,
           "Infer torch script model inferer failed, output type is not supported. output type: " << output.tagKind());
      throw InfererException("Infer torch script model inferer failed, output type is not supported. output type: " +
                             output.tagKind());
    }
  } catch (const c10::Error& e) {
    LOG4(ERROR, "Infer torch script model inferer failed, error: " << e.what());
    throw InfererException(e.what());
  } catch (const std::exception& e) {
    LOG4(ERROR, "Infer torch script model inferer failed, error: " << e.what());
    throw InfererException(e.what());
  }
}

void TorchModelInferer::BatchInfer(const std::vector<std::pair<std::string, TensorWrapper>>& inputs,
                                   std::vector<std::pair<std::string, TensorWrapper>>& outputs,
                                   std::vector<GrpsContext*>& ctxs) {
  Infer(inputs, outputs, *ctxs[0]);
}

void TorchModelInferer::InferWithProfiler(const std::string& profiler_path,
                                          const std::vector<std::pair<std::string, TensorWrapper>>& inputs,
                                          std::vector<std::pair<std::string, TensorWrapper>>& outputs,
                                          GrpsContext& context) {
  throw InfererException("Not supported now.");
}

void TorchModelInferer::BatchInferWithProfiler(const std::string& profiler_path,
                                               const std::vector<std::pair<std::string, TensorWrapper>>& inputs,
                                               std::vector<std::pair<std::string, TensorWrapper>>& outputs,
                                               std::vector<GrpsContext*>& ctxs) {
  throw InfererException("Not supported now.");
}

// --------------------------------------- No converter mode [BEGIN] ---------------------------------------

void TorchModelInferer::Infer(const ::grps::protos::v1::GrpsMessage& inputs,
                              ::grps::protos::v1::GrpsMessage& outputs,
                              GrpsContext& context) {
  std::vector<std::pair<std::string, TensorWrapper>> input_tensors;
  std::vector<std::pair<std::string, TensorWrapper>> output_tensors;
  converter_.PreProcess(inputs, input_tensors, context);
  Infer(input_tensors, output_tensors, context);
  converter_.PostProcess(output_tensors, outputs, context);
}

void TorchModelInferer::InferWithProfiler(const std::string& profiler_path,
                                          const ::grps::protos::v1::GrpsMessage& inputs,
                                          ::grps::protos::v1::GrpsMessage& outputs,
                                          GrpsContext& context) {
  throw InfererException("Not supported now.");
}

void TorchModelInferer::BatchInfer(std::vector<const ::grps::protos::v1::GrpsMessage*>& inputs,
                                   std::vector<::grps::protos::v1::GrpsMessage*>& outputs,
                                   std::vector<netease::grps::GrpsContext*>& ctxs) {
  std::vector<std::pair<std::string, TensorWrapper>> input_tensors;
  std::vector<std::pair<std::string, TensorWrapper>> output_tensors;
  converter_.BatchPreProcess(inputs, input_tensors, ctxs);
  BatchInfer(input_tensors, output_tensors, ctxs);
  converter_.BatchPostProcess(output_tensors, outputs, ctxs);
}

void TorchModelInferer::BatchInferWithProfiler(const std::string& profiler_path,
                                               std::vector<const ::grps::protos::v1::GrpsMessage*>& inputs,
                                               std::vector<::grps::protos::v1::GrpsMessage*>& outputs,
                                               std::vector<netease::grps::GrpsContext*>& ctxs) {
  throw InfererException("Not supported now.");
}

// --------------------------------------- No converter mode [END] ---------------------------------------

} // namespace netease::grps